pkg/controller/clone/planner.go

```diff
@@ -0,0 +1,534 @@
+}
+
+func (p *Planner) watchOwned(log logr.Logger, obj client.Object) error {
+	objList := p.RootObjectType.DeepCopyObject().(client.ObjectList)
+	if err := p.Controller.Watch(&source.Kind{Type: obj}, handler.EnqueueRequestsFromMapFunc(
+		func(obj client.Object) (reqs []reconcile.Request) {
+			uid, ok := obj.GetLabels()[p.OwnershipLabel]
+			if !ok {
+				return
+			}
+			matchingFields := client.MatchingFields{
+				p.UIDField: uid,
+			}
+			if err := p.Client.List(context.Background(), objList, matchingFields); err != nil {
+				log.Error(err, "Unable to list resource", "matchingFields", matchingFields)
+				return
+			}
+			sv := reflect.ValueOf(objList).Elem()
+			iv := sv.FieldByName("Items")
+			for i := 0; i < iv.Len(); i++ {
+				o := iv.Index(i).Addr().Interface().(client.Object)
+				reqs = append(reqs, reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Namespace: o.GetNamespace(),
+						Name:      o.GetName(),
+					},
+				})
+			}
+			return
+		}),
+	); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Planner) computeStrategyForSourcePVC(ctx context.Context, args *ChooseStrategyArgs) (*cdiv1.CDICloneStrategy, error) {
+	if ok, err := p.validateTargetStorageClassAssignment(ctx, args); !ok || err != nil {
+		return nil, err
+	}
+
+	sourceClaim := &corev1.PersistentVolumeClaim{}
+	exists, err := getResource(ctx, p.Client, args.DataSource.Namespace, args.DataSource.Spec.Source.Name, sourceClaim)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		message := fmt.Sprintf(MessageCloneWithoutSource, "pvc", args.DataSource.Spec.Source.Name)
+		p.Recorder.Event(args.TargetClaim, corev1.EventTypeWarning, CloneWithoutSource, message)
+		args.Log.V(3).Info("Source PVC does not exist, cannot compute strategy")
+		return nil, nil
+	}
+
+	if err = p.validateSourcePVC(args, sourceClaim); err != nil {
+		p.Recorder.Event(args.TargetClaim, corev1.EventTypeWarning, CloneValidationFailed, MessageCloneValidationFailed)
+		args.Log.V(1).Info("Validation failed", "target", args.TargetClaim, "source", sourceClaim)
+		return nil, err
+	}
+
+	strategy := cdiv1.CloneStrategySnapshot
+	cs, err := GetGlobalCloneStrategyOverride(ctx, p.Client)
+	if err != nil {
+		return nil, err
+	}
+
+	if cs != nil {
+		strategy = *cs
+	} else if args.TargetClaim.Spec.StorageClassName != nil {
+		sp := &cdiv1.StorageProfile{}
+		exists, err := getResource(ctx, p.Client, "", *args.TargetClaim.Spec.StorageClassName, sp)
+		if err != nil {
+			return nil, err
+		}
+
+		if !exists {
+			args.Log.V(3).Info("missing storageprofile for", "name", *args.TargetClaim.Spec.StorageClassName)
+		}
+
+		if exists && sp.Status.CloneStrategy != nil {
+			strategy = *sp.Status.CloneStrategy
+		}
+	}
+
+	if strategy == cdiv1.CloneStrategySnapshot {
+		n, err := GetCompatibleVolumeSnapshotClass(ctx, p.Client, sourceClaim, args.TargetClaim)
+		if err != nil {
+			return nil, err
+		}
+
+		if n == nil {
+			strategy = cdiv1.CloneStrategyHostAssisted
```

Comment:
Can we emit an event on the target claim here that explains we switched to host assisted clone because no compatible volume snapshot class could be found?